### PR TITLE
fix: sync email_verified from auth0 on demand

### PR DESCRIPTION
**Pull request by jzsfkzm**
refs #1863

## Summary by CodeRabbit

* **New Features**
  * Added "Verify Email" API endpoint, website verify-email page with success/failure feedback and auto-continue, client SDK verifyEmail(email) method, and a timed AutoButton UI component.

* **Security**
  * Verification enforces that only the signed-in user may verify their own email; verification status is synced from the identity provider.

* **Integration**
  * Added identity provider integration to fetch remote email verification status and sync with local user records.

* **Tests**
  * New unit and functional tests covering verify-email behaviors and edge cases.

### CODE REVIEW

File: apps/api/src/auth/routes/email-verified/email-verified.router.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+const EmailVerifiedQuerySchema = z.object({
+  email: z.string()
+});
+
+export type EmailVerifiedQuery = z.infer<typeof EmailVerifiedQuerySchema>;
+
+const route = createRoute({
+  method: "get",
+  path: "/v1/email-verified",
```
[the remark this answers is not in the document]
Comment (jzsfkzm) [reply apps/api/src/auth/routes/email-verified/email-verified.router.ts]: Auth0 can send the user's browser to a "redirect to" URL after verifying email. This will be that URL, that's why it's a GET endpoint.

File: apps/api/src/auth/routes/email-verified/email-verified.router.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
+import { createRoute } from "@hono/zod-openapi";
+import { container } from "tsyringe";
+import { z } from "zod";
+
+import { AuthController } from "@src/auth/controllers/auth/auth.controller";
+import { OpenApiHonoHandler } from "@src/core/services/open-api-hono-handler/open-api-hono-handler";
+
+export const emailVerifiedRouter = new OpenApiHonoHandler();
```
[the remark this answers is not in the document]
Comment (jzsfkzm) [reply apps/api/src/auth/routes/email-verified/email-verified.router.ts]: Moved schema to a schema file, no circular references any more.

File: apps/api/src/auth/routes/email-verified/email-verified.router.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    301: {
+      description: "Redirects to the front page"
+    }
+  }
+});
+
+emailVerifiedRouter.get(route.path, async function emailVerified(c) {
+  await container.resolve(AuthController).syncEmailVerified({ email: c.req.query("email") as string });
+  return c.redirect("/", 302);
+});
```
[the remark this answers is not in the document]
Comment (jzsfkzm) [reply apps/api/src/auth/routes/email-verified/email-verified.router.ts]: I'd rather keep `emailVerifiedRouter.get`, so the endpoint won't land in Swagger docs.

File: apps/api/src/auth/routes/email-verified/email-verified.router.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+const route = createRoute({
+  method: "get",
+  path: "/v1/email-verified",
+  summary: "Checks if the email is verified",
+  tags: ["Users"],
+  request: {
+    query: EmailVerifiedQuerySchema
+  },
+  responses: {
+    301: {
```
Comment (baktun14) [inline apps/api/src/auth/routes/email-verified/email-verified.router.ts]:
```suggestion
    302: {
```

File: apps/api/src/services/db/userDataService.ts [hunk 5]
```diff
@@ -221,3 +221,16 @@ export async function subscribeToNewsletter(userId: string) {
     }
   );
 }
+
+export async function setEmailVerified(userId: string, emailVerified: boolean) {
```
Comment (baktun14) [inline apps/api/src/services/db/userDataService.ts]: Let's move this in the `UserService`

**Comment by baktun14** [comment]
Next step will be to implement the page and then we can update the email template?

File: apps/api/test/functional/auth0.spec.ts [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+import { UserRepository } from "@src/user/repositories/user/user.repository";
+
+import { UserSeeder } from "@test/seeders/user.seeder";
+
+jest.setTimeout(30000);
+
+describe("Auth0 Email Verification", () => {
+  afterEach(() => {
+    jest.clearAllMocks();
+  });
```
Comment (ygrishajev) [inline apps/api/test/functional/auth0.spec.ts]: one of the point of using `setup` function is that jest hooks aren't needed as every test case gets it's own clean state

File: apps/api/src/auth/controllers/auth/auth.controller.ts [hunk 7]
```diff
@@ -20,4 +24,27 @@ export class AuthController {
       await this.auth0.sendVerificationEmail(currentUser.userId);
     }
   }
+
+  @Protected()
+  async syncEmailVerified({ data: { email } }: VerifyEmailRequest) {
+    const { currentUser } = this.authService;
+
+    const targetEmail = email.trim().toLowerCase();
+    const currentEmail = currentUser.email?.trim().toLowerCase();
```
Comment (ygrishajev) [inline apps/api/src/auth/controllers/auth/auth.controller.ts]: this duplicates zod schema transfomations

File: apps/api/src/auth/controllers/auth/auth.controller.ts [hunk 8]
```diff
@@ -20,4 +24,27 @@ export class AuthController {
...
+    assert(currentUser.userId && currentEmail === targetEmail, 401);
+
+    const auth0User = await this.auth0.getUserByEmail(targetEmail);
+    assert(auth0User, 404);
+
+    await this.userService.upsertUser({
+      userId: currentUser.userId,
+      emailVerified: auth0User.email_verified
+    });
```
Comment (ygrishajev) [inline apps/api/src/auth/controllers/auth/auth.controller.ts]: I think that it's a bit too much logic for a controller. It seems a bit more appropriate to implement a certain method inside user service (`syncEmailVerified`), it would be more re-usable

File: apps/api/src/auth/controllers/auth/auth.controller.ts [hunk 9]
```diff
@@ -20,4 +24,27 @@ export class AuthController {
...
     }
   }
+
+  @Protected()
+  async syncEmailVerified({ data: { email } }: VerifyEmailRequest) {
+    const { currentUser } = this.authService;
+
+    const targetEmail = email.trim().toLowerCase();
+    const currentEmail = currentUser.email?.trim().toLowerCase();
+    assert(currentUser.userId && currentEmail === targetEmail, 401);
```
Comment (ygrishajev) [inline apps/api/src/auth/controllers/auth/auth.controller.ts]: 403 code here seems more appropriate

Comment (ygrishajev) [reply apps/api/src/auth/controllers/auth/auth.controller.ts]: This looks like a permission job here. `REGULAR_*` roles can be updated to contain permissions like

```
{ action: ["read", "update"], subject: "User", conditions: { $or: [{ id: "${user.id}" }, { email: "${user.email}" }] } },
```
or
```
{ action: "verify-email", subject: "User", conditions: { email: "${user.email}" } },
```

and then checked against permissions:
```
ability.can("update", subject("User", { email })
ability.can("verify-email", subject("User", { email })
```

we are already planning admin dashboard where straight permission check can be problematic.

File: apps/api/src/auth/controllers/auth/auth.controller.ts (hunk 7 above)
Comment (jzsfkzm) [reply apps/api/src/auth/controllers/auth/auth.controller.ts]: Oh, indeed, turning email into targetEmail was unnecessary as we already transform that input with the schema.

**Review by ygrishajev — COMMENTED** [review]
thanks for the changes @jzsfkzm! i think it's essential to update the type here so it's not misused accidentally https://github.com/akash-network/console/pull/1880/files#diff-98e087b6c1c3c2f322904604e83ce6cb3632af0db4438fc354d08945d9428d37R55 - should also have optional email

File: apps/api/test/functional/verify-email.spec.ts [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+    });
+
+    async function setup(input: { emailVerified: boolean }) {
+      const walletTestingService = new WalletTestingService(app);
+      const { user, token } = await walletTestingService.createRegisteredUser({
+        email_verified: input.emailVerified
+      });
+
+      const auth0Service = container.resolve(Auth0Service);
+      jest.spyOn(auth0Service, "getUserByEmail").mockImplementation(async (email: string) => {
```
Comment (ygrishajev) [inline apps/api/test/functional/verify-email.spec.ts]: I'm thinking this should be mocked via request interception as this is a functional? wdyt @baktun14 @stalniy ?

File: apps/api/test/functional/verify-email.spec.ts [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+
+      expect(response.status).toBe(200);
+      const result = await response.json();
+      expect(result).toMatchObject({
+        data: {
+          emailVerified: true
+        }
+      });
+
+      const userSetting = await UserSetting.findOne({ where: { userId: user.userId } });
```
Comment (ygrishajev) [inline apps/api/test/functional/verify-email.spec.ts]: as this is a functional test, I'm thinking this should be using endpoints.

```ts
const currentUserResponse = await app.request("/v1/user/me", {
  method: "GET",
  headers: {
    authorization: `Bearer ${token}`,
    "content-type": "application/json"
  }
});
const body = await currentUserResponse.json();
expect(body.data.emailVerified).toBe(true);
```

File: apps/api/test/functional/verify-email.spec.ts [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+      });
+
+      const userAuthTokenService = container.resolve(UserAuthTokenService);
+      jest.spyOn(userAuthTokenService, "getValidUserId").mockResolvedValue(user.userId);
+
+      const userRepository = container.resolve(UserRepository);
+      jest.spyOn(userRepository, "findById").mockResolvedValue(user);
+      jest.spyOn(userRepository, "findByUserId").mockResolvedValue(user);
+
+      jest.spyOn(UserSetting, "update");
```
Comment (ygrishajev) [inline apps/api/test/functional/verify-email.spec.ts]: I don't think any mocks above are needed (besides 3rd party auth0). App should function normally without tweaks

File: apps/deploy-web/src/pages/user/verify-email/index.tsx [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+        <h3 className="text-2xl">
+          {isVerifying && "Just a moment while we finish verifying your email."}
+          {isVerified === true && "Your email was verified. You can continue using the application."}
+          {isVerified === false && "Your email was not verified. Please try again."}
+        </h3>
+
+        {isVerified === true && (
+          <div className="pt-6">
+            <Link href={UrlService.home()} className={cn(buttonVariants({ variant: "default" }), "inline-flex items-center")}>
+              Continue
```
Comment (ygrishajev) [inline apps/deploy-web/src/pages/user/verify-email/index.tsx]: shouldn't we just navigate and show some success message?

File: packages/http-sdk/src/auth/auth-http.types.ts [hunk 14]
```diff
@@ -0,0 +1,9 @@
+import { z } from "zod";
+
+export const VerifyEmailResponseSchema = z.object({
```
Comment (ygrishajev) [inline packages/http-sdk/src/auth/auth-http.types.ts:3]: why is schema needed here? perhaps just type would be enough?

File: apps/deploy-web/src/pages/user/verify-email/index.tsx [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+    if (email) {
+      verifyEmail(email);
+    }
+  });
+
+  return (
+    <Layout isLoading={isVerifying}>
+      <NextSeo title="Verifying your email" />
+
+      <div className="mt-10 text-center">
```
Comment (baktun14) [inline apps/deploy-web/src/pages/user/verify-email/index.tsx]: We can add the Akash Console logo and big icon (check or X or spinner) to make it prettier and more interactive.

File: apps/deploy-web/src/pages/user/verify-email/index.tsx (hunk 13 above)
Comment (baktun14) [reply apps/deploy-web/src/pages/user/verify-email/index.tsx]: I agree, we could add a little coutndown "5 seconds until being redirected" and keep the continue button to fast forward

**Comment by jzsfkzm** [comment]
> thanks for the changes @jzsfkzm! i think it's essential to update the type here so it's not misused accidentally https://github.com/akash-network/console/pull/1880/files#diff-98e087b6c1c3c2f322904604e83ce6cb3632af0db4438fc354d08945d9428d37R55 - should also have optional email

Hm, that's right, types should be the same, but shouldn't it rather be accepting the same UserOutput we're passing when using this method?
```diff
-  getAbilityFor(role: Role, user: { userId: string | null }) {
+  getAbilityFor(role: Role, user: UserOutput) {
```

**Comment by ygrishajev** [comment]
> Hm, that's right, types should be the same, but shouldn't it rather be accepting the same UserOutput we're passing when using this method?

this can work too

File: apps/api/test/functional/verify-email.spec.ts [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+
+        return Promise.resolve(null);
+      });
+
+      const userAuthTokenService = container.resolve(UserAuthTokenService);
+      jest.spyOn(userAuthTokenService, "getValidUserId").mockResolvedValue(user.userId);
+
+      const userRepository = container.resolve(UserRepository);
+      jest.spyOn(userRepository, "findById").mockResolvedValue(user);
+      jest.spyOn(userRepository, "findByUserId").mockResolvedValue(user);
```
Comment (ygrishajev) [inline apps/api/test/functional/verify-email.spec.ts]: none of these mocks (except 3rd part auth0) is really needed i think. it should be possible to create a user and it would work without them

File: apps/api/src/auth/controllers/auth/auth.controller.ts [hunk 17]
```diff
@@ -20,4 +23,16 @@ export class AuthController {
       await this.auth0.sendVerificationEmail(currentUser.userId);
     }
   }
+
+  @Protected()
```
Comment (baktun14) [inline apps/api/src/auth/controllers/auth/auth.controller.ts:27]:
```suggestion
  @Protected([{ action: "verify-email", subject: "User" }])
```

Comment (baktun14) [reply apps/api/src/auth/controllers/auth/auth.controller.ts:27]: Instead of the `this.authService.throwUnlessCan("verify-email", "User", { email });`

Comment (jzsfkzm) [reply apps/api/src/auth/controllers/auth/auth.controller.ts:27]: But it is not the same, is it?

`@Protected([{ action: "verify-email", subject: "User" }])` means the user acting can `verify-email` some `User` in general. We don't check if they can `verify-email` the User in request (identified by the email address we send).

On the other hand, `this.authService.throwUnlessCan("verify-email", "User", { email })` would be ok only for the user with the same email, and superusers (who have `manage/all`).

What do you think?

Comment (baktun14) [reply apps/api/src/auth/controllers/auth/auth.controller.ts:27]: Ahh I see you're right nvm

File: apps/deploy-web/src/pages/user/verify-email/index.tsx (hunk 13 above)
Comment (jzsfkzm) [reply apps/deploy-web/src/pages/user/verify-email/index.tsx]: Added such a timeout, now it will redirect automatically.

File: apps/api/src/auth/services/auth0/auth0.service.spec.ts [hunk 18]
```diff
@@ -0,0 +1,142 @@
...
+        });
+      }
+
+      const auth0Service = new (class extends Auth0Service {
+        constructor() {
+          super(mockAuthConfig);
+          // Override the managementClient property
+          Object.assign(this, { managementClient: mockManagementClient });
+        }
+      })();
```
Comment (ygrishajev) [inline apps/api/src/auth/services/auth0/auth0.service.spec.ts]: this is the reason to extract ManagementClient to a dedicated provider and inject it into the auth0service

File: apps/api/src/auth/services/auth0/auth0.service.spec.ts [hunk 19]
```diff
@@ -0,0 +1,142 @@
...
+    it("should propagate errors from Auth0 API", async () => {
+      const email = faker.internet.email();
+      const error = new Error("Auth0 API error");
+
+      const { auth0Service, mockGetByEmail } = setup({
+        mockError: error
+      });
+
+      await expect(auth0Service.getUserByEmail(email)).rejects.toThrow("Auth0 API error");
+      expect(mockGetByEmail).toHaveBeenCalledWith({ email });
```
Comment (ygrishajev) [inline apps/api/src/auth/services/auth0/auth0.service.spec.ts]: the code doesn't handle errors and thus should not test them

File: apps/api/test/functional/verify-email.spec.ts [hunk 20]
```diff
@@ -0,0 +1,161 @@
...
+  });
+
+  async function setup(input: { emailVerified: boolean }) {
+    const walletTestingService = new WalletTestingService(app);
+    const { user, token } = await walletTestingService.createRegisteredUser({
+      email_verified: input.emailVerified
+    });
+
+    const auth0Service = container.resolve(Auth0Service);
+    jest.spyOn(auth0Service, "getUserByEmail").mockImplementation(async (email: string) => {
```
Comment (ygrishajev) [inline apps/api/test/functional/verify-email.spec.ts]: should be mocked via request interception or on sdk level (ManagementClient)

File: apps/api/src/auth/services/auth0/auth0.service.spec.ts [hunk 21]
```diff
@@ -0,0 +1,142 @@
...
+    it("should return user when user is found", async () => {
+      const email = faker.internet.email();
+      const mockUser: Partial<GetUsers200ResponseOneOfInner> = {
+        user_id: faker.string.uuid(),
+        email: email,
+        email_verified: true,
+        name: faker.person.fullName(),
+        created_at: faker.date.past().toISOString(),
+        updated_at: faker.date.recent().toISOString()
+      };
```
Comment (ygrishajev) [inline apps/api/src/auth/services/auth0/auth0.service.spec.ts]: this is prob reused in here and in functional tests, should be extracted to a seeder.

File: apps/api/src/auth/services/auth0/auth0.service.spec.ts (hunk 18 above)
Comment (jzsfkzm) [reply apps/api/src/auth/services/auth0/auth0.service.spec.ts]: Add a provider for ManagementClient.

File: apps/api/src/auth/services/auth0/auth0.service.spec.ts (hunk 19 above)
Comment (jzsfkzm) [reply apps/api/src/auth/services/auth0/auth0.service.spec.ts]: Indeed, this class is opaque for errors, dropped this test.

File: apps/api/test/functional/verify-email.spec.ts (hunk 20 above)
Comment (jzsfkzm) [reply apps/api/test/functional/verify-email.spec.ts]: Started mocking on the ManagementClient level. I chose this as we don't really know (and I think we're fine with not knowing) about how it works on the HTTP level.

File: apps/api/src/auth/services/auth0/auth0.service.spec.ts (hunk 21 above)
Comment (jzsfkzm) [reply apps/api/src/auth/services/auth0/auth0.service.spec.ts]: That's right, added a seeder for Auth0 users.

File: apps/api/src/core/providers/auth.provider.ts [hunk 22]
```diff
@@ -0,0 +1,26 @@
...
+        clientSecret: authConfig.get("AUTH0_M2M_SECRET")
+      });
+    }
+
+    return ManagementClientFactory.instance;
+  }
+}
+
+container.register(ManagementClient, {
+  useFactory: () => ManagementClientFactory.getInstance()
```
Comment (ygrishajev) [inline apps/api/src/core/providers/auth.provider.ts]: this is marked a singleton higher so the container should make sure this would be a singleton.

basically this should work:

```typescript
import { ManagementClient } from "auth0";
import { container } from "tsyringe";

import { AuthConfigService } from "@src/auth/services/auth-config/auth-config.service";

export const AUTH0_MANAGEMENT_CLIENT = "AUTH0_MANAGEMENT_CLIENT";

container.register(AUTH0_MANAGEMENT_CLIENT, {
  useFactory: c => {
    const config = c.resolve(AuthConfigService);
    return new ManagementClient({
      domain: config.get("AUTH0_M2M_DOMAIN"),
      clientId: config.get("AUTH0_M2M_CLIENT_ID"),
      clientSecret: config.get("AUTH0_M2M_SECRET")
    });
  }
});


class Foo () {
   constructor(@inject(AUTH0_MANAGEMENT_CLIENT) private readonly managementClient: ManagementClient) {}
}
```

or 



```typescript
import { ManagementClient } from "auth0";
import { container } from "tsyringe";

import { AuthConfigService } from "@src/auth/services/auth-config/auth-config.service";

container.register(ManagementClient, {
  useFactory: c => {
    const config = c.resolve(AuthConfigService);
    return new ManagementClient({
      domain: config.get("AUTH0_M2M_DOMAIN"),
      clientId: config.get("AUTH0_M2M_CLIENT_ID"),
      clientSecret: config.get("AUTH0_M2M_SECRET")
    });
  }
});

// above needs to be imported somewhere in the root of the app

class Foo () {
   constructor(private readonly managementClient: ManagementClient) {}
}
```

Comment (jzsfkzm) [reply apps/api/src/core/providers/auth.provider.ts]: Simplified it a lot now, but the instance injected still has to be a singleton - otherwise the instance used in production and the one we get from DI in the test will be different instances, so when we mock the latter, the prior will just go unmocked, and cause errors, and tests will break this way.

**Review by ygrishajev — APPROVED** [review]

File: apps/api/src/core/providers/auth.provider.ts [hunk 23]
```diff
@@ -0,0 +1,22 @@
+import { ManagementClient } from "auth0";
+import { container } from "tsyringe";
+
+import { AuthConfigService } from "@src/auth/services/auth-config/auth-config.service";
+
+let managementClientInstance: ManagementClient | null = null;
```
Comment (ygrishajev) [inline apps/api/src/core/providers/auth.provider.ts:6]: non-blocking: why is this important to make it a singleton? 

if it is needed indeed, found this helper https://github.com/Microsoft/tsyringe?tab=readme-ov-file#instancecachingfactory
